File: VB/WindowsFormsApplication1/MyTreeList.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes
Imports System.ComponentModel

Namespace WindowsFormsApplication1
	  <System.ComponentModel.DesignerCategory("")> _
	  Public Class InitNewNodeEventArgs
		  Inherits EventArgs
		Public Sub New(ByVal nodeId As Integer)
			Me.nodeId_Renamed = nodeId
		End Sub
		Private nodeId_Renamed As Integer

		Public ReadOnly Property NodeId() As Integer
			Get
				Return nodeId_Renamed
			End Get
		End Property
	  End Class


	Public Class MyTreeList
		Inherits TreeList
		Public Delegate Sub InitNewRowEventHandler(ByVal sender As Object, ByVal a As InitNewNodeEventArgs)
		Public  Event InitNewNode As InitNewRowEventHandler
		Private _NewAddedNode As TreeListNode

		Public Sub New()
			MyBase.New()

		End Sub

		Public Sub New(ByVal ignore As Object)
			MyBase.New(ignore)
		End Sub



		Public Sub AddNewNode()
			Dim Id As Integer = -1

			_NewAddedNode = Me.AppendNode(Nothing, Id)
			isNewRowDirty = False
			Me.FocusedNode = _NewAddedNode
			RaiseInitNewNode(New InitNewNodeEventArgs(_NewAddedNode.Id))
		End Sub

		Private isNewRowDirty As Boolean = False
		Protected Overrides Sub OnMouseDown(ByVal e As MouseEventArgs)
			Dim info As TreeListHitInfo = Me.CalcHitInfo(e.Location)
			If info.HitInfoType = HitInfoType.Cell Then
					isNewRowDirty = True
					Me.RefreshNode(_NewAddedNode)
			End If
			MyBase.OnMouseDown(e)
		End Sub


		Protected Overrides Sub RaiseCustomDrawNodeIndicator(ByVal e As CustomDrawNodeIndicatorEventArgs)
			If e.Node IsNot _NewAddedNode OrElse isNewRowDirty Then
				MyBase.RaiseCustomDrawNodeIndicator(e)
				Return
			End If

			e.ImageIndex = 1
			e.Painter.DrawObject(e.ObjectArgs)

		End Sub

		Protected Overridable Sub RaiseInitNewNode(ByVal e As InitNewNodeEventArgs)
            RaiseEvent InitNewNode (Me,e)
		End Sub

		Public Function IsNewNode(ByVal nodeId As Integer) As Boolean
			Return nodeId = _NewAddedNode.Id
		End Function


	End Class
End Namespace

File: CS/WindowsFormsApplication1/MyTreeList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;
using System.ComponentModel;

namespace WindowsFormsApplication1 {
    [System.ComponentModel.DesignerCategory("")]
    public class InitNewNodeEventArgs : EventArgs {
        public InitNewNodeEventArgs(int nodeId) {
            this.nodeId = nodeId;
        }
        private int nodeId;
       
        public int NodeId {
            get { return nodeId; }
        }
    }

 
    public class MyTreeList : TreeList {
        public delegate void InitNewNodeEventHandler(object sender, InitNewNodeEventArgs a);
        private event InitNewNodeEventHandler initNewNode;
        private TreeListNode _NewAddedNode;

        public MyTreeList()
            : base() {
               
        }

        [ Category("Data")]
		public event InitNewNodeEventHandler InitNewNode {
			add { this.Events.AddHandler(initNewNode, value); }
			remove { this.Events.RemoveHandler(initNewNode, value); }
		}
        public MyTreeList(object ignore) : base(ignore) { }
     
      

        public void AddNewNode() {
            int Id = -1;

            _NewAddedNode = this.AppendNode(null, Id);
            isNewRowDirty = false;
            this.FocusedNode = _NewAddedNode;
            RaiseInitNewNode(new InitNewNodeEventArgs(_NewAddedNode.Id));
        }

        private bool isNewRowDirty = false;
        protected override void OnMouseDown(MouseEventArgs e) {
            TreeListHitInfo info = this.CalcHitInfo(e.Location);
            if (info.HitInfoType == HitInfoType.Cell) {
                    isNewRowDirty = true;
                    this.RefreshNode(_NewAddedNode);
            }
            base.OnMouseDown(e);
        }

        
        protected override void RaiseCustomDrawNodeIndicator(CustomDrawNodeIndicatorEventArgs e) {
            if (e.Node != _NewAddedNode || isNewRowDirty) {
                base.RaiseCustomDrawNodeIndicator(e);
                return;
            }

            e.ImageIndex = 1;
            e.Painter.DrawObject(e.ObjectArgs);
          
        }

        protected virtual void RaiseInitNewNode(InitNewNodeEventArgs e) {
            InitNewNodeEventHandler handler = (InitNewNodeEventHandler)this.Events[initNewNode];

            if (handler != null) {
                handler(this, e);
            }
        }

        public bool IsNewNode(int nodeId) {
            return nodeId == _NewAddedNode.Id;
        }

        
    }
}
